xayn-ai/src/coi/point.rs

```diff
@@ -50,6 +53,26 @@ impl Default for CoiStats {
     }
 }
 
+#[derive(Clone, Deserialize, PartialEq, Serialize)]
+#[cfg_attr(test, derive(Debug))]
+pub(crate) struct KeyPhrase {
+    pub(super) words: String,
+    pub(super) point: Embedding,
+}
+
+impl KeyPhrase {
+    pub(super) fn is_valid(&self) -> bool {
+        !self.words.is_empty() && self.point.iter().copied().all(f32::is_finite)
+    }
+
+    pub(super) fn is_unique(key_phrases: &[KeyPhrase]) -> bool {
+        key_phrases
```

Comment:
no we can't, as explained in the pr description:

> the uniqueness invariant can't be enforced for example by a HashSet because KeyPhrase can't implement Eq + Hash and also not by sort & dedup because KeyPhrase can't implement PartialOrd/Ord

this is because of the limitation of the underlying `ArrayBase`.

Reply:
if we can assume that two `KeyPhrase` with the same `words` then also the embedding is the same we can implement `Eq` and `Hash` ignoring the embedding.

Reply:
unfortunately we can't assume that. for example `"key phrase"` and `"Key Phrase"` may or may not produce different embeddings, e.g. in cased bert models. so we would have to rely on `SMBert` never changing, otherwise we have errors in here without noticing.

Reply:
changed as discussed.